### PR TITLE
lp:1722789 Intermediary slave does not log master changes with binlog…

### PR DESCRIPTION
**Pull request by vlad-lesin**
…_rows_query_log_events

Let's consider the following function:

int ha_blackhole::update_row(const uchar *old_data, uchar *new_data)
{
  DBUG_ENTER("ha_blackhole::update_row");
  THD *thd= ha_thd();
  if (is_slave_applier(thd) && thd->query().str == NULL)
    DBUG_RETURN(0);
  DBUG_RETURN(HA_ERR_WRONG_COMMAND);
}

In the case if this function is invoked from slave thread and it's
query string is empty, then it's supposed the function finishes successfully,
otherwise some error is returned. The same logic exists in some other blackhole
engine functions.

If "binlog_rows_query_log_events = 1" binary log contains events with queries.
The queries are ignored, but during "Rows_query_log_event" event execution
(see Rows_query_log_event::do_apply_event()) thread query string is filled with
the query from the event. And when then Update_rows_log_event::do_exec_row()
is executed, sql thread has query in thd->m_query_string, and this condition:

  is_slave_applier(thd) && thd->query().str == NULL

becomes false, what leads to errors during execution of blackhole engine
functions such as ha_blackhole::update_row(). Such errors are output to error
log and sql thread stops, as a result is't binary log is not filling.

The question is what does this
"is_slave_applier(thd) && thd->query().str == NULL" condition mean? I suppose
the condition means that the currently processed relay log event type is not
QUERY_EVENT.

That is why  I would propose the following fix:
replace "thd->query().str == NULL" with
"thd->rli_slave->rows_query_ev || thd->query() == NULL".

See also: https://github.com/percona/percona-server/pull/1998.
Testing: https://jenkins.percona.com/job/percona-server-5.6-param/2026/

### CODE REVIEW

File: storage/blackhole/ha_blackhole.cc [hunk 1]
```diff
@@ -120,7 +121,8 @@ int ha_blackhole::update_row(const uchar *old_data, uchar *new_data)
 {
   DBUG_ENTER("ha_blackhole::update_row");
   THD *thd= ha_thd();
-  if (is_slave_applier(thd) && thd->query() == NULL)
+  if (is_slave_applier(thd) &&
```
Comment (laurynas-biveinis) [inline storage/blackhole/ha_blackhole.cc]: Please extract this condition to a helper method

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]
FWIW, the original condition might have been added for this (commit message of 43fb8633b0774f00dce6490fd93deebdfe4ab386):

"    It is necessary to only pretend this for the SQL thread, since
    a SELECT executed on the Blackhole engine will otherwise never
    return EOF, causing a livelock."

But it is not clear why it is not enough to check the thread type only

File: storage/blackhole/ha_blackhole.cc [hunk 2]
```diff
@@ -20,6 +20,7 @@
 #include "probes_mysql.h"
 #include "ha_blackhole.h"
 #include "sql_class.h"                          // THD, SYSTEM_THREAD_SLAVE_*
+#include "rpl_rli.h"
```
Comment (laurynas-biveinis) [inline storage/blackhole/ha_blackhole.cc]: Also please add a comment ```// THD::rli_slave::rows_query_ev```

File: storage/blackhole/ha_blackhole.cc (hunk 1 above)
Comment (vlad-lesin) [reply storage/blackhole/ha_blackhole.cc]: fixed

File: storage/blackhole/ha_blackhole.cc (hunk 2 above)
Comment (vlad-lesin) [reply storage/blackhole/ha_blackhole.cc]: fixed

**Comment by vlad-lesin** [comment]
Ready for the next round.

**Review by laurynas-biveinis — APPROVED** [review]
Please also send this upstream

**Pull request merged**